File: Day1-9/1.py
def part1(nums):
    d = {}
    for n in lines:
        d[n] = True

    for n in lines:
        if 2020 - n not in d:
            continue
        return n * (2020 - n)


def part2(nums):
    d = {}
    for n1 in nums:
        for n2 in nums:
            d[2020 - (n1 + n2)] = (n1, n2)

    for n in nums:
        if n not in d:
            continue
        n1, n2 = d[n]
        return n1 * n2 * n


if __name__ == "__main__":
    with open("1.txt") as f:
        lines = [int(l) for l in f.readlines()]
    p1 = part1(lines)
    print("Part 1: {}".format(p1))
    p2 = part2(lines)
    print("Part 2: {}".format(p2))

File: Day1-9/1.hs
import           Data.Map   (Map)
import qualified Data.Map   as Map
import           Data.Maybe (catMaybes)

part1 :: [Integer] -> Integer
part1 nums = match * (2020 - match)
  where
    lookup = Map.fromList $ map (\x -> (x, x)) nums
    match = head $ catMaybes $ map (\x -> Map.lookup (2020 - x) lookup) nums

part2 :: [Integer] -> Integer
part2 nums = a * b * (2020 - a - b)
  where
    lookup =
      Map.fromList $
      map (\([x, y]) -> ((2020 - x - y, (x, y)))) $ sequence [nums, nums]
    (a, b) = head $ catMaybes $ map (\x -> Map.lookup x lookup) nums

main = do
  content <- readFile "1.txt"
  let inputLines = lines content
  let nums = map (\x -> read x :: Integer) inputLines
  putStrLn $ "Part 1: " ++ show (part1 nums)
  putStrLn $ "Part 2: " ++ show (part2 nums)
